zcash_client_backend/src/data_api/chain.rs

```diff
@@ -236,46 +239,47 @@ where
 /// # Ok(())
 /// # }
 /// ```
-pub fn scan_cached_blocks<E, N, P, C, D>(
+pub fn scan_cached_blocks<'a, E, N, P, C, D>(
     params: &P,
     cache: &C,
-    data: &mut D,
+    data: Arc<Mutex<D>>,
     limit: Option<u32>,
-) -> Result<(), E>
+) -> Box<dyn Future<Item = (), Error = E> + Send + 'a>
 where
-    P: consensus::Parameters,
+    P: consensus::Parameters + Send + Sync,
     C: BlockSource<Error = E>,
     D: WalletWrite<Error = E, NoteRef = N>,
-    N: Copy + Debug,
-    E: From<Error<N>>,
+    N: Copy + Debug + Send,
+    E: From<Error<N>> + Send + 'a,
 {
-    let sapling_activation_height = params
+    let mut data_guard = data.lock().unwrap();
+    let sapling_activation_height = try_f!(params
         .activation_height(NetworkUpgrade::Sapling)
-        .ok_or(Error::SaplingNotActive)?;
+        .ok_or(Error::SaplingNotActive));
 
     // Recall where we synced up to previously.
     // If we have never synced, use sapling activation height to select all cached CompactBlocks.
-    let mut last_height = data.block_height_extrema().map(|opt| {
+    let mut last_height = try_f!(data_guard.block_height_extrema().map(|opt| {
         opt.map(|(_, max)| max)
             .unwrap_or(sapling_activation_height - 1)
-    })?;
+    }));
```

Comment:
Returning the `Future` doesn't make fn async/non-blocking automatically. It still performs a blocking IO (at least in SQLite client case), by calling `block_height_extrema` and other `WalletRead/Write` methods. In case of a long-running query, it may block the executor thread for some time, preventing other futures from being executed.

Also, we will still need to implement `WalletRead/Write` trait for `WalletDbInner` in our DeFi framework. Will it cause problems as forementioned traits are not async, while almost all IndexedDB operations are? What is our plan on this?

Reply:
great question! thank you.
for the first note, I guess refactoring the functions to async will fix the issue outlined but that will require a lot of changes to the library.

Yes, I'm aware from the start that this will also be a big problem as most `indexeddb` operations rely heavily on async operations so I thought to just focus on `blockdb` blocking issues in this PR and when I start implementing `WalletRead/Write` trait for `WalletDbInner`, then I'll make recommended changes to this lib as it will be very important in having `walletdb` in `wasm`

Reply:
Is there a way to make as fewer changes to the original lib code as possible? It would be best if original code is not changed at all. I'm asking it because the more changes we make, the harder it will be to sync with upstream (which already went very far ahead). Possible mistakes during merging/conflict solving are especially dangerous in this case, because `librustzcash` deals with users' funds directly.

cc @shamardy

Reply:
> Is there a way to make as fewer changes to the original lib code as possible? It would be best if original code is not changed at all.

Yes, we can wrap the original, unchanged code in an async or future returning function, e.g
```rust
pub fn with_blocks_fut<F>(
    cache: &BlockDb,
    from_height: BlockHeight,
    limit: Option<u32>,
    mut with_row: Box<F>,
) -> Box<dyn Future<Item = (), Error = SqliteClientError> + Send>
    where
        F: FnMut(CompactBlock) -> Result<(), SqliteClientError>,
{
    let res = try_f!(with_blocks(cache, from_height, limit, *with_row));
    Box::new(futures01::future::ok(res))
}
```
We can even make this new code part of a feature if these new functiosn are in the same files as the original ones as suggest here https://github.com/KomodoPlatform/librustzcash/pull/5#issuecomment-1612390823 but it's not necessary. 

But for blocking IO functions such as `block_height_extrema` we would need 2 new async functions one for native (sqlite) target that wraps the original and uses `spawn_blocking` or `async_blocking` after we port it from komoDeFi and the other one will be async for wasm IndexedDB target. So I think creating new separate crates that wrap these functions might be a better solution than including them in the same files. What do you think @borngraced @artemii235 ?

Reply:
> We can even make this new code part of a feature if these new functiosn are in the same files as the original ones as suggest here https://github.com/KomodoPlatform/librustzcash/pull/5#issuecomment-1612390823 but it's not necessary.

Yeah I suppose this works In this case.

> But for blocking IO functions such as block_height_extrema we would need 2 new async functions one for native (sqlite) target that wraps the original and uses spawn_blocking or async_blocking after we port it from komoDeFi and the other one will be async for wasm IndexedDB target. So I think creating new separate crates that wrap these functions might be a better solution than including them in the same files. What do you think @borngraced @artemii235 ?

either this or modifying the lib directly will still requires a whole lot changes as walletdb has gotten so many I/O functions that will need to be ported as well

Reply:
> So I think creating new separate crates that wrap these functions might be a better solution than including them in the same files.

Yeah, or maybe as separate modules in the lib to avoid adding one more crate to the dependency tree.